### DISCUSSION TITLE
pygetpapers

### DISCUSSION THREAD

**Discussion by petermr** [petermr/petermr#10] [Announcements]
# pygetpapers

Automatic downloading of scientific articles, owner: @ayush4921

Posting of
news, updates, problems, suggestions for `pygetpapers`.

**Comment by petermr** [comment]
# automated subject-based scoring for `pygetpapers`

(Will ask Daniel Mietchen also to contribute)

## Opportunity and goals

Current literature searches are based on text in:
### bibliographic info (authors, journal, date, publisher, and citations)
This is generally only of interest to academics and similar researchers but is hugely influenced by academic ranking tools

### references
Probably mainly of use for citation analysis
 
### openness and licences
Important but highly variable and poorly supported.

### titles. 
Very common but often crude and limited (authors pick "clever" titles and they are very general.

### abstracts
More useful but often paywalled. Some subjects (e.g. trials) have structured abstracts but the structure may not be used.

### fulltext
Extremely valuable. (we'll concentrate on this)

### tables
Also extremely valuable. The captions often indicate clearly the disciplines used in the paper.

### image captions
Also very valuable and similar to tables.

### image content
Requires specialist software but very powerful in some disciplines (e.g. chemical structures, statistics plots, flowcharts). Can be split into:
* **image retrieval**. "This paper contains phylogenetic trees"
* **image analysis**.  "we extracted the chemical formulae of 12 natural products from this paper".

For EuropePMC content we can analyse all of these automatically. For rxivs it depends on whether the HTML is monolithic or the PDF is easily parsable For open content / publically visible on websites we have to scrape (may cause problems). 

More details and proposal in `reply`

**Comment by petermr** [comment]
# automated subject-based metadata and scoring

## goal
* To have `pygetpapers` (or linked software). calculate novel (subject-based) metadata (working title `pyg-meta`) and scoring for hits it retrieves.
* To store `pyg-meta` in Wikidata pages (or linkable from/to Wikidata) for each paper.
* To incorporate `pyg-meta`  in a `pyg` score for each hit
* To offer the reader search filtering/prioritization based on the `pyg-meta`

## Information Retrieval (IR-based) metadata
This is content/metadata whose presence can be easily determined/calculated. It general includes Boolean (is this info present?) and summary counts ("8 tables")

It could include:
* language(s) used. Critical to support non-EN content.
* number of tables (sometimes not paywalled)
* number of figures
* length of abstracts
* structure of abstracts (are there sections? were patients involved?). 
* boilerplates (CoI(ConflictsOfInterest), ethics, contributors, funders, data availability, etc.) IR gives Boolean for presence.
* links (supplemental data, etc.)

## Information Extraction (IE-based) metadata

This is information extracted from content (text, tables, figures/images). It can be generic (e.g. SOLR/Lucene for words in text - very common, and probably the basis of most conventional search tools) or specific (e.g. OSCAR for chemical names and interpretation). It generally involves specialist tools. There are hundreds of disciplines with well developed metadata, e.g.:
* gene names
* nucleic and protein sequences
* chemical formulae
* statistics (e.g. Statcheck)
* trial data (human, animals, societal, etc.) (subject count, organizations, trial-ID, duration, protocols,...)
* organisms (especially genera and binomial names)
* research locations (often different from author locations!)
* instruments used (very powerful, as they normally have well defined IDs in the text)
* reagents used (similarly)
* methodology (often systematized by name)
* research programs by name (e.g. geoscience, astronomy, environment)
* geopolitical (organizations, laws, etc.)

`pyg` can support many of these through `ami` dictionaries.

# fulltext and metadata indexing 

Propose that we start with indexing fulltext in repositories (initially EPMC but later the `rxiv`s). We decide on a core set of IE processes that are widely useful, high precision (few false positives), easy to implement and could be offered to Wikimedia with a good chance of being acceptable. The indexing could either run autonomously independently of queries to `pygetpapers` or could be manually fed to Wikimedia. A Wikimedia cloud which both carries out the indexing and stores the results  sounds great. 

## novel metadata (indexed into Wikidata)
(This is "fairly easy" to index). The results are highly standardized and "factual".
(Please add more details, if you are confident they are "important" and "easy" to implement).
The discipline stuff could produce main_content if backed with frequencies. Searches could be Boolean (is_found) or quantitative (frequencies).

### sectioning
we assume the paper is a least somewhat sectionable (e.g. bibliodata,  abstract, body, references, supplement). Then any indexing should indicate where it's located. 

### bibliography
(I omit the author names, journal, references, citations, etc.)
* author countries
* length of paper
* table count
* figure count
* openness and licences

### metacontent (requires fulltext)
* funders
* CoI
* ethics
* datasets and availability
* code availability

### body content (generic)
* country_count
* organizations

### body text content (discipline) -
* binomial species and genera (style/regex/WD_lookup)
* diseases (WD-based dictionary)
* viruses (WD-based dictionary)
* chemicals (WD-based dictionary) OR regex/OSCAR
* bioreagents (WD-based dictionary)
* statistics methods (WD-based dictionary)

### figure captions and table captions
Easy to extract and analyse in the same way

### ML and NLP-based searches (lower priority)
* software names
* Hearst patterns
...

# Methodology

## architecture
### conventional
* **client** runs `pygetpapers` locally
* papers stored on client disk
*  results analysed with `docanalysis` and `py4ami` ("`ami`") and further tools

### simple cloud-based
This saves compute-time locally and download traffic
* client sends queries to cloud
* pygetpapers downloads hits and papers to cloud (no traffic to client)
* `ami` runs in cloud and computes metadata and scores
* optional : papers are preprocessed (e.g. annotated or transformed)
* client selects scored papers and downloads
 

### cached cloud
* pygetpapers sends query to repo
* hits sent to ami-cache in cloud
* if cache contains paper it's added to hitlist with score
* else paper downloaded to cache and score computed and cached
* client selects papers

### wikidata main subject
* pygetpapers sends query to repo
* hits searched against wikidata main subject
* score computed 
* client selects papers

### wikidata pyg-meta (this is the most exciting if WD agrees)
* wikidata includes a pyg-meta score (Qitem)
* pygetpapers sends query to repo
* hits searched against wikidata-pyg-meta
* if present added to scores
* else paper downloaded to cloud, pyg-meta computed and added to wikidata
* client selects papers

**Comment by ayush4921** [comment]
It seems like a great metric to add to scientific literature. I am happy to contribute to the scripting and design of cloud infrastructure. I do have a few concerns though: 

1) How do we determine if research should be scored? Should all research be scored? And should the metrics remain regardless of the domain they fall under? 

2)Are there costs associated with running the cloud? How much storage is needed/necessary and how much do we access?

3)Is human moderation required for the scoring, or is it meant to be fully autonomous?

**Comment by petermr** [comment]
On Mon, Jul 25, 2022 at 5:55 AM Ayush Garg ***@***.***> wrote:

> It seems like a great metric to add to scientific literature. I am happy
> to contribute to the scripting and design of cloud infrastructure. I do
> have a few concerns though:
>
>    1. How do we determine if research should be scored? Should all
>    research be scored? And should the metrics remain regardless of the domain
>    they fall under?
>
> The metrics are not to  evaluate the research but to provide search
metadata. . They should be independent of domain. The *user* can them to
decide whether to download the paper.

>
>
> 2)Are there costs associated with running the cloud? How much storage is
> needed/necessary and how much do we access?
>
Daniel says that Wikidata provides free storage and cycles if it the
purpose is Wikidata.

> Daniel
>
3)Is human moderation required for the scoring, or is it meant to be fully
> autonomous?
>

Fully automatic. We publish the algorithms/parameters and expose the
software.

> —
> Reply to this email directly, view it on GitHub
> <https://github.com/petermr/petermr/discussions/10#discussioncomment-3218830>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFTCS4FPWY725VG3WTYG4LVVYM4BANCNFSM5LRZHOBQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Peter Murray-Rust
Founder ContentMine.org
and
Reader Emeritus in Molecular Informatics
Dept. Of Chemistry, University of Cambridge, CB2 1EW, UK

**Comment by EmanuelFaria** [comment]
This paper got me thinking about other ranking criteria that may a be useful after the basics are in place and working...

**Retrospective studies – utility and caveats**
doi: 10.4997/JRCPE.2020.409


[Retrospective studies – utility and caveats.pdf](https://github.com/petermr/petermr/files/9220517/Retrospective.studies.utility.and.caveats.pdf)

**Comment by petermr** [comment]
Thanks
There are zillions of meta-analyses - not sure that they are relevant to us.


On Fri, Jul 29, 2022 at 2:55 PM Emanuel Faria ***@***.***>
wrote:

> This paper got me thinking about other ranking criteria that may a be
> useful after the basics are in place and working...
>
> *Retrospective studies – utility and caveats*
> doi: 10.4997/JRCPE.2020.409
>
> Retrospective studies – utility and caveats.pdf
> <https://github.com/petermr/petermr/files/9220517/Retrospective.studies.utility.and.caveats.pdf>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/petermr/petermr/discussions/10#discussioncomment-3280890>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFTCS26RBHAA7NLDPW3ZQ3VWPPFZANCNFSM5LRZHOBQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Peter Murray-Rust
Founder ContentMine.org
and
Reader Emeritus in Molecular Informatics
Dept. Of Chemistry, University of Cambridge, CB2 1EW, UK
